src/brogue/Items.c

```diff
@@ -3536,59 +3545,121 @@ void negate(creature *monst) {
+            }
+            if (monst->status[STATUS_LEVITATING] > 1 ) {
+                monst->status[STATUS_LEVITATING] = 1;
+                negated = true;
+            }
             if (monst->status[STATUS_DARKNESS]) {
                 monst->status[STATUS_DARKNESS] = 0;
                 updateMinersLightRadius();
                 updateVision(true);
+                negated = true;
             }
         } else {
-            monst->status[STATUS_TELEPATHIC] = 0;
-            monst->status[STATUS_MAGICAL_FEAR] = 0;
-            monst->status[STATUS_LEVITATING] = 0;
+            if (monst->status[STATUS_TELEPATHIC] > 0 ) {
+                monst->status[STATUS_TELEPATHIC] = 0;
+                negated = true;
+            }
+            if (monst->status[STATUS_MAGICAL_FEAR] > 0 ) {
+                monst->status[STATUS_MAGICAL_FEAR] = 0;
+                negated = true;
+            }
+            if (monst->status[STATUS_LEVITATING] > 0 ) {
+                monst->status[STATUS_LEVITATING] = 0;
+                negated = true;
+            }
         }
-        monst->info.flags &= ~MONST_IMMUNE_TO_FIRE;
-        monst->movementSpeed = monst->info.movementSpeed;
-        monst->attackSpeed = monst->info.attackSpeed;
+        if (monst->info.flags & MONST_IMMUNE_TO_FIRE) {
+            monst->info.flags &= ~MONST_IMMUNE_TO_FIRE;
+            negated = true;
+        }
+        if (monst->movementSpeed != monst->info.movementSpeed) {
+            monst->movementSpeed = monst->info.movementSpeed;
+            negated = true;
+        }
+        if (monst->attackSpeed != monst->info.attackSpeed) {
+            monst->attackSpeed = monst->info.attackSpeed;
+            negated = true;
+        }
+
         if (monst != &player && monst->mutationIndex > -1 && mutationCatalog[monst->mutationIndex].canBeNegated
             && rogue.patchVersion >= 3) {
 
             monst->mutationIndex = -1;
+            negated = true;
         }
         if (monst != &player && (monst->info.flags & NEGATABLE_TRAITS)) {
             if ((monst->info.flags & MONST_FIERY) && monst->status[STATUS_BURNING]) {
                 extinguishFireOnCreature(monst);
             }
             monst->info.flags &= ~NEGATABLE_TRAITS;
+            negated = true;
             refreshDungeonCell(monst->xLoc, monst->yLoc);
             refreshSideBar(-1, -1, false);
         }
         for (i = 0; i < 20; i++) {
-            backupBolts[i] = monst->info.bolts[i];
-            monst->info.bolts[i] = BOLT_NONE;
-        }
-        for (i = 0, j = 0; i < 20 && backupBolts[i]; i++) {
-            if (boltCatalog[backupBolts[i]].flags & BF_NOT_NEGATABLE) {
-                monst->info.bolts[j] = backupBolts[i];
-                j++;
+            if (monst->info.bolts[i]) {
+                if (!(boltCatalog[monst->info.bolts[i]].flags & BF_NOT_NEGATABLE)) {
+                    monst->info.bolts[i] = BOLT_NONE;
+                    negated = true;
+                }
```

Comment:
Is this really equivalent to before? Previously the loop was collecting the bolts to the start, but now there may be BOLT_NONE gaps in the list, which breaks other looping I believe

Reply:
I think it's equivalent and doesn't break anything but will double-check.

Reply:
To be safe, I reverted this refactor and added the negation detection to the existing loop.

Reply:
It's possible I missed something, but I didn't see code which kept track of two list indices, which are necessary for compacting the non-negated bolts into a contiguous list, which is required as bolt iteration in the code seems to go up to the first none bolt. Thanks for the fix in any case